src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,212 @@
+    }
+
+    internal static bool CanRewrite(BoundDelegateCreationExpression boundDelegateCreation)
+    {
+        var targetMethod = boundDelegateCreation.MethodOpt;
+
+        Debug.Assert(targetMethod is { });
+
+        return targetMethod.IsStatic && !boundDelegateCreation.IsExtensionMethod;
+    }
+
+    internal BoundExpression Rewrite(BoundDelegateCreationExpression boundDelegateCreation)
+    {
+        var targetMethod = boundDelegateCreation.MethodOpt;
+        var delegateType = boundDelegateCreation.Type;
+
+        Debug.Assert(targetMethod is { });
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = boundDelegateCreation.Syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(receiver: null, cacheField);
+        var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        // We don't need to synthesize a container for each and every function.
+        //
+        // For example:
+        //   void LF1<T>()
+        //   {
+        //       void LF2<G>()
+        //       {
+        //           void LF3()
+        //           {
+        //               Func<T> d = SomeMethod<T>;
+        //               static void LF4 () { Func<T> d = SomeMethod<T>; }
+        //           }
+        //           
+        //           void LF5()
+        //           {
+        //               Func<T> d = SomeMethod<T>;
+        //           }
+        //       }
+        //   }
+        //
+        // In the above case, only one cached delegate is necessary, and it could be assigned to the container 'owned' by LF1.
+
+        if (!TryGetOwnerFunction(_factory.CurrentFunction, delegateType, targetMethod, out var ownerFunction))
+        {
+            var typeCompilationState = _factory.CompilationState;
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new DelegateCacheContainer(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containers = _genericCacheContainers ??= new Dictionary<MethodSymbol, DelegateCacheContainer>();
```

Comment:
Do we need a custom equality comparer here?

Reply:
> Do we need a custom equality comparer here?

It looks like the keys here are definitions, therefore, they will not be different only due to insignificant things. We could use reference identity comparer here though. More efficient than the default one.

Reply:
We already have a type - ```ReferenceEqualityComparer```

Reply:
Great! Thanks. Added in aaa2621, please check.